activerecord/lib/active_record/relation/batches.rb

```diff
@@ -222,45 +222,53 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
       end
 
       relation = relation.reorder(batch_order(order)).limit(batch_limit)
-      relation = apply_limits(relation, start, finish, order)
-      relation.skip_query_cache! # Retaining the results in the query cache would undermine the point of batching
+      relation = apply_finish_limit(relation, finish, order) if finish
       batch_relation = relation
 
       loop do
+        batch_relation = apply_start_limit(relation, start, order) if start
+
         if load
-          records = batch_relation.records
-          ids = records.map(&:id)
-          yielded_relation = where(primary_key => ids)
-          yielded_relation.load_records(records)
+          records = batch_relation.uncached do
+            batch_relation.limit(batch_limit + 1).records
+          end
+
+          start = records[batch_limit]&.id
+          records = records.take(batch_limit)
+
+          break if records.empty?
+
+          raise ArgumentError.new("Primary key not included in the custom select clause") unless records.first.id
+
+          batch_relation.load_records(records)
         else
-          ids = batch_relation.pluck(primary_key)
-          yielded_relation = where(primary_key => ids)
-        end
+          stop = batch_relation.uncached do
+            batch_relation.offset(batch_limit).pick(primary_key)
```

Comment:
So, if you do `offset(batch_limit -1)` you:

  - Scan 1 less record (not a huge deal but might as well)
  - No longer need the `inclusive: false` option.

I think that could simplify the code a bit

Reply:
On an a batch edges, we need to have in a previous batch `id < stop` and in the following `id >= stop`. At the beginning of the loop we have `batch_relation = apply_start_limit(relation, start, order) if start` (which does `>=`), so we need to do `<` here and need to get a `batch_limit + 1`th record.
If we will do `offset(batch_limit - 1)`, we will have a batch of smaller size by 1.